torch/nn/modules/module.py

```diff
@@ -254,32 +257,38 @@ def register_forward_hook(self, hook):
         return handle
 
     def __call__(self, *input, **kwargs):
-        for hook in self._forward_pre_hooks.values():
-            hook(self, input)
-        result = self.forward(*input, **kwargs)
-        for hook in self._forward_hooks.values():
-            hook_result = hook(self, input, result)
-            if hook_result is not None:
-                raise RuntimeError(
-                    "forward hooks should never return any values, but '{}'"
-                    "didn't return None".format(hook))
-        if len(self._backward_hooks) > 0:
-            var = result
-            while not isinstance(var, Variable):
-                if isinstance(var, dict):
-                    var = next((v for v in var.values() if isinstance(v, Variable)))
-                else:
-                    var = var[0]
-            grad_fn = var.grad_fn
-            if grad_fn is not None:
-                for hook in self._backward_hooks.values():
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        with torch.jit.scope('%s$%s' % (self.__class__.__name__, self._name),
+                             tuple(torch.autograd.function._iter_variables(input))):
+            for hook in self._forward_pre_hooks.values():
+                hook(self, input)
+            result = self.forward(*input, **kwargs)
+            for hook in self._forward_hooks.values():
+                hook_result = hook(self, input, result)
+                if hook_result is not None:
+                    raise RuntimeError(
+                        "forward hooks should never return any values, but '{}'"
+                        "didn't return None".format(hook))
+            if len(self._backward_hooks) > 0:
+                var = result
+                while not isinstance(var, Variable):
+                    if isinstance(var, dict):
+                        var = next((v for v in var.values() if isinstance(v, Variable)))
+                    else:
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in self._backward_hooks.values():
+                        wrapper = functools.partial(hook, self)
+                        functools.update_wrapper(wrapper, hook)
+                        grad_fn.register_hook(wrapper)
         return result
 
     def __setstate__(self, state):
         self.__dict__.update(state)
+        if '_name' not in self.__dict__:
+            self._name = ''
+            for name, module in self.named_children():
+                module._name = name
```

Comment:
`__setstate__` will also be called for children, so this unnecessarily makes loading the tree O(n^2) instead of O(n)

Reply:
So, just to make sure I get it right: `__setstate__` will be called once per module, and each module will iterate its _immediate_ children and set their names. Wouldn't this operation only touch each module once, not multiple times?

Reply:
@apaszke I didn't make changes here, but please let me know if I'm getting it wrong.

Reply:
Yes, this looks fine to me.

Reply:
My bad. The complexity is correct, but the code isn't. If you have e.g. 4 levels of nesting like this, the names will end up as:
```
m1            ''
  m2          'm2'
    m3        ''     # NOTE: m2 had _name set, so it didn't iterate
      m4      'm4' 
```

There's still a bit of a problem with models that are DAGs - see my comment below

Reply:
Nice catch! Of course it wouldn't work.
This
```
if '_name' not in self.__dict__:
    self._name = ''
for name, module in self.named_children():
    if '_name' not in module.__dict__:
        module._name = name
```
would fix it, at the cost of checking all children even if `name_` is in the parent.

However, thinking about DAG, it may make no sense to assign a name attribute to a module itself (which is why it wasn't there in the first place 😄, I get it now). 

I'm realizing that the way to get the name for a module (to auto-generate the scope name for the module) is to obtain it from the stack you're talking about in your DAG comment.

Reply:
Yup, the stack is a reliable way to get this, without all the `_name` maintenance headaches.